### PR TITLE
Feature: add loading indicator when ReconnectApp is running

### PR DESCRIPTION
**Pull request by nkdengineer**
### Fixed Issues

$ https://github.com/Expensify/App/issues/46611
PROPOSAL: https://github.com/Expensify/App/issues/46611#issuecomment-2262387280

### Tests

1. Login to app
2. Go to offline mode
3. Back to online mode
4. Verify that: loading indicator is displayed
- [x] Verify that no errors appear in the JS console

### QA Steps

1. Login to app
2. Go to offline mode
3. Back to online mode
4. Verify that: loading indicator is displayed
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/fe95c11e-3f0a-466e-819d-01e610e32c9e

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/d2450106-a797-4907-9a1b-07aa1990c01b

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/67293301-8b07-4b8a-adb0-e6866357e5c2

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/fb4e02f8-a9e0-46af-82b0-55934a87a0a6

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/302ea093-ca05-4b90-b966-6e3c36355106

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/5fc1e299-db5b-4186-91f7-62d44ac33bbf

</details>

### CODE REVIEW

**Comment by dubielzyk-expensify** [comment]
We should be using our color variables here so on light mode we should end up with this:
CleanShot 2024-09-09 at 19 41 53@2x

It looks like the background of the loading indicator is a dark mode color?

**Comment by nkdengineer** [comment]
@dubielzyk-expensify I updated to fix this case.

**Comment by dubielzyk-expensify** [comment]
Awesome. Can you update the recordings please 😇?

**Comment by nkdengineer** [comment]
@dubielzyk-expensify i'll update soon

**Comment by nkdengineer** [comment]
@dubielzyk-expensify @getusha i updated, please check again

**Comment by dannymcclain** [comment]
I am really excited for this to be in the product 😊

**Comment by dubielzyk-expensify** [comment]
This is looking awesome. Tiny tiny nitpick, but do the background of the loading bar and the loading bar itself have rounded corners? If so, can we remove them? It looks a bit better when this is square to match the page header border

CleanShot 2024-09-10 at 11 09 03@2x
CleanShot 2024-09-10 at 11 09 59@2x

File: src/components/ProgressBar.tsx
Comment (getusha) [inline src/components/ProgressBar.tsx:1]: @nkdengineer lets make a constant variable for every number we use for the animation with descriptive names.

File: src/components/ProgressBar.tsx
Comment (getusha) [inline src/components/ProgressBar.tsx:28]: should we add this to const as well?

File: src/components/ProgressBar.tsx
Comment (getusha) [inline src/components/ProgressBar.tsx:26]: same here

**Review by getusha — COMMENTED** [review]
minor comments

File: src/components/ProgressBar.tsx
Comment (nkdengineer) [reply src/components/ProgressBar.tsx:26]: @getusha We didn't add it in the const with the parameters of Easing.bezier like [here](https://github.com/Expensify/App/blob/e52d43c2cd8ae7d8449161651867e8b163ac1321/src/components/AutoCompleteSuggestions/BaseAutoCompleteSuggestions.tsx#L69)

Comment (nkdengineer) [reply src/components/ProgressBar.tsx:26]: friendly bump @getusha

**Comment by getusha** [comment]
@nkdengineer the UI is still jumpy, didn't we position it absolute?

https://github.com/user-attachments/assets/c83a369e-542d-4227-895f-b6fbb703cdaf

**Comment by nkdengineer** [comment]
> @nkdengineer the UI is still jumpy, didn't we position it absolute?

@getusha I fixed, please check again

File: src/components/ProgressBar.tsx
Comment (getusha) [inline src/components/ProgressBar.tsx:66]: Any way to hide the container? i can see it with a white background when it isn't active

Screenshot 2024-10-01 at 12 24 36 in the afternoon

**Comment by dubielzyk-expensify** [comment]
This is looking great! All 👍 from me on the visual end. Love the entry and exit fade too

**Comment by getusha** [comment]
Not sure if this is related but the performance on native is really bad it's nearly unusable, i'll check if this is any different on main


https://github.com/user-attachments/assets/460b389e-0c1d-4852-a7c3-8a97155d43ee

**Comment by blimpich** [comment]
@getusha did you get a chance to test this on main? Maybe we just need to update this branch with the latest changes on main?

File: src/components/ProgressBar.tsx
Comment (nkdengineer) [reply src/components/ProgressBar.tsx:66]: From this [comment](https://github.com/Expensify/App/pull/48772#issuecomment-2385348856), Do I need to fix this?

Comment (getusha) [reply src/components/ProgressBar.tsx:66]: I think so yes, looks off to leave it like that. WDYT @dubielzyk-expensify?

**Comment by getusha** [comment]
I'll test this again and complete the checklist by the end of the day. the performance is fine on main.

**Comment by getusha** [comment]
@nkdengineer could you investigate this? i can't open a chat. the app eventually crashes.

https://github.com/user-attachments/assets/c1748b27-b573-452f-9698-68ce97059622

**Comment by nkdengineer** [comment]
@getusha i can't reproduce this bug. Can you add more details steps here?

**Comment by nkdengineer** [comment]
@getusha friendly bump

**Comment by nkdengineer** [comment]
@getusha Just merged.

**Comment by getusha** [comment]
Still the same, could you try reproducing it again @nkdengineer?

https://github.com/user-attachments/assets/2e273b9d-191c-4936-86d9-bef25d3c8013

**Comment by nkdengineer** [comment]
@getusha Work for me. Did you run IOS again or only run metro to open the IOS app?

https://github.com/user-attachments/assets/58cea4f6-8e0a-4c60-8b37-971ec94aebfb

**Comment by srikarparsi** [comment]
Hm, I'm having some problems with my ios emulator but will try to get it figured out. In the meanwhile, were you able to test if the same behavior is occurring for you on main?

**Comment by srikarparsi** [comment]
I'm not able to replicate what you're seeing @getusha. I'm going to trigger an Ad Hoc build and we can go from there:

https://github.com/user-attachments/assets/9fa9d7a7-c994-4a90-9bcb-e4eb1a1bcaf6

**Comment by srikarparsi** [comment]
AdHoc: https://github.com/Expensify/App/actions/runs/11409038457

**Comment by srikarparsi** [comment]
I'm not experiencing this [problem](https://github.com/Expensify/App/pull/48772#issuecomment-2385367747) on native iOS. @getusha could you please check if you have a chance and if not we can merge this

**Comment by srikarparsi** [comment]
Hey @getusha, we have a limited number of devices we can add and we're quite on the edge so I asked @mountiny if he could test. He saw some weird behavior so I wanted to ask, @nkdengineer, can you please merge main and I'll generate a new build and test?

**Comment by nkdengineer** [comment]
@srikarparsi merged main, please move forward

**Comment by shawnborton** [comment]
Hmm I keep getting this on the test build, any ideas?
CleanShot 2024-10-22 at 08 41 38@2x

**Comment by srikarparsi** [comment]
Yeah, it looks like we introduced a problem with login which is being looked at [here](https://github.com/Expensify/App/issues/51222). ([Slack thread](https://expensify.slack.com/archives/C049HHMV9SM/p1729583811580549?thread_ts=1729574128.664629&cid=C049HHMV9SM)). Once it's fixed, we'll have to merge main again and spin up a new AdHoc build. Once this [PR](https://github.com/Expensify/App/pull/51226) gets merged I'll move forward with that.

**Comment by srikarparsi** [comment]
In the meanwhile @shawnborton you can login with a new account (that hasn't been created before) and should be able to test it out.

**Comment by srikarparsi** [comment]
Hi @nkdengineer, do you think you could merge main one more time as the sign in bug got merged [here](https://github.com/Expensify/App/pull/51226/files)?

**Comment by mountiny** [comment]
IMG_7139
works fine for me, but the indicator is on the Report page too, I assume that should not be the case, right?

**Comment by dannymcclain** [comment]
>works fine for me, but the indicator is on the Report page too, I assume that should not be the case, right?

I believe this is expected if ReconnectApp is indeed still running.

**Comment by nkdengineer** [comment]
@srikarparsi I merged.

**Comment by mountiny** [comment]
> I believe this is expected if ReconnectApp is indeed still running.

I am not sure, I feel like its misleading as the OpenReport is what is loading all the details for the report so if anything I would expect the loader to indicate that the OpenReport is running on that screen.

**Comment by srikarparsi** [comment]
Yeah, I'm not really sure but kind of agree with what @dannymcclain said [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1719238100079219?thread_ts=1699972596.342559&cid=C01GTK53T8Q)

I also generated a new AdHoc build https://github.com/Expensify/App/actions/runs/11490278696

**Comment by mountiny** [comment]
No problem, I do not want to block on this, I just wanted to note that it was kind of confusing for me. Essentially you are on the Report screen, you see that _something_ is loading but when it stops, nothing really changes for you (as it was not loading the specific report data necessarily)

**Comment by dannymcclain** [comment]
>No problem, I do not want to block on this, I just wanted to note that it was kind of confusing for me. Essentially you are on the Report screen, you see that something is loading but when it stops, nothing really changes for you (as it was not loading the specific report data necessarily)

Oh dang. Maybe I have been misunderstanding this from the start. So are you saying that ReconnectApp may or _may not_ be loading data for the specific report you're on? In your comment you said nothing really changed for you, but is it possible that something might change based on ReconnectApp running/finishing? Just trying to figure out if it makes sense to keep it on the report screen or not. If it's NEVER going to impact what you see on a report then maybe it's not... But if it _might_, then it probably is.

Would love some more thoughts from @Expensify/design. I'm worried that I haven't quite understood what ReconnectApp is actually doing all this time 😅

(I also don't want to block on this because I think this is a hugely valuable improvement to the app—but we can always make follow-ups if we need to)

**Comment by shawnborton** [comment]
> you see that something is loading but when it stops, nothing really changes for you (as it was not loading the specific report data necessarily)

I guess I would view this as the app is checking to see if anything has changed, and it very well could be that nothing has changed but the app needs to think about it and check first. So I don't find this too weird personally I guess? But I also don't feel too strongly, even just having this on the LHN only is a big improvement.

**Comment by dannymcclain** [comment]
>I guess I would view this as the app is checking to see if anything has changed, and it very well could be that nothing has changed but the app needs to think about it and check first. So I don't find this too weird personally I guess? 

Yeah that makes sense to me.

>But I also don't feel too strongly, even just having this on the LHN only is a big improvement.

In the interest of getting this in the product so we can see how it feels, I'd be down to just proceed how it's been implemented thus far. If we find that it DOES feel super strange, we could come back to it?

**Comment by dubielzyk-expensify** [comment]
Yeah I don't have much to add other than that I agree with all viewpoints here. I guess the only thing is that I think we should get this into the product, cause I think that we need to see how this performs in real life scenario and over some time before making adjustments. I'm happy to do LHN only though, but agree with Shawn that it's more of a "check in" and that things changing or not isn't really that important

**Comment by srikarparsi** [comment]
Cool, @getusha do you think you can finish reviewing this whenever you have a chance? If iOS native is still performing slow for you, I think we can merge without it because I just tested on my device with this [build](https://github.com/Expensify/App/pull/48772#issuecomment-2433976689) and it seems to be working fine.

**Comment by srikarparsi** [comment]
Friendly bump on the above @getusha, also there's conflicts now @nkdengineer

**Comment by nkdengineer** [comment]
Resolved conflict

File: src/components/ProgressBar.tsx
Comment (nkdengineer) [reply src/components/ProgressBar.tsx:66]: @getusha i fixed

**Comment by nkdengineer** [comment]
@getusha i fixed

**Comment by nkdengineer** [comment]
@getusha i updated

**Review by getusha — APPROVED** [review]

**Review by getusha — APPROVED** [review]

File: src/components/LoadingBar.tsx
Comment (srikarparsi) [inline src/components/LoadingBar.tsx:59]: changes (with an s)

File: src/components/LoadingBar.tsx
Comment (srikarparsi) [inline src/components/LoadingBar.tsx:7]: Can this be a regular comment (// instead of /**) Also, you can remove the "with fade out animation". So the comment should be "Whether or not to show the loading bar"

**Review by srikarparsi — CHANGES_REQUESTED** [review]
Just two small changes to the comments but other than that looks good to me

**Comment by nkdengineer** [comment]
@srikarparsi i updated

**Review by srikarparsi — APPROVED** [review]
Awesome thanks!

**Comment by Julesssss** [comment]
Hey @srikarparsi @nkdengineer & @getusha. This PR appears to cause HybridApp to crash.  We're getting close to deploying App so I might need to revert this. 

Screenshot 2024-11-05 at 14 43 01

**Comment by staszekscp** [comment]
We've just discovered that bumping `react-native-reanimated` to 3.16 should fix it, @blazejkustra is currently working on the bump!

**Comment by Julesssss** [comment]
Excellent, thanks!

**Comment by Julesssss** [comment]
Being fixed [here](https://github.com/Expensify/App/issues/52053), no need to revert

**Pull request merged**
